File: 2023/rust/src/bin/d08p1.rs
use std::{collections::HashMap, fs};

type Map<'a> = HashMap<&'a str, Node<'a>>;

#[derive(Debug)]
enum Direction {
    Left,
    Right,
}
impl From<char> for Direction {
    fn from(value: char) -> Self {
        match value {
            'L' => Direction::Left,
            'R' => Direction::Right,
            _ => panic!("Unknown direction"),
        }
    }
}

#[derive(Debug)]
struct Node<'a> {
    name: &'a str,
    left: &'a str,
    right: &'a str,
}
impl<'a> Node<'a> {
    fn from_str(s: &'a str) -> Result<Self, ()> {
        let (name, (left, right)) = s
            .split_once(" = ")
            .and_then(|s| {
                Some((
                    s.0,
                    s.1.strip_prefix('(')
                        .and_then(|s| s.strip_suffix(')'))
                        .and_then(|s| s.split_once(", "))?,
                ))
            })
            .ok_or(())?;

        Ok(Node { name, left, right })
    }
}

fn main() {
    println!("Day 8 - Part 1");

    let str = fs::read_to_string("../_input/day08.txt").unwrap();
    let lines: Vec<&str> = str.lines().collect();

    let total = process(lines);
    println!("Total: {}", total);
}

fn process(lines: Vec<&str>) -> u32 {
    let instructions: Vec<Direction> = lines[0].chars().map(|i| i.into()).collect();
    let map = process_map(&lines[2..lines.len()]);

    let mut current_node = "AAA";
    let mut current_dir = 0;
    let mut hit = false;
    let mut counter = 0;
    while !hit {
        let node = map.get(current_node).unwrap();
        match instructions[current_dir] {
            Direction::Left => current_node = node.left,
            Direction::Right => current_node = node.right,
        }

        counter += 1;
        if current_node.eq("ZZZ") {
            hit = true;
        }

        current_dir += 1;
        if current_dir >= instructions.len() {
            current_dir = 0;
        }
    }

    counter
}

fn process_map<'a>(lines: &'a [&'a str]) -> Map<'a> {
    let mut map = Map::new();

    for s in lines {
        let node = Node::from_str(s).unwrap();
        map.insert(node.name, node);
    }

    map
}

#[cfg(test)]
mod test {
    use crate::process;

    #[test]
    fn test() {
        assert_eq!(
            process(
                r"RL

AAA = (BBB, CCC)
BBB = (DDD, EEE)
CCC = (ZZZ, GGG)
DDD = (DDD, DDD)
EEE = (EEE, EEE)
GGG = (GGG, GGG)
ZZZ = (ZZZ, ZZZ)"
                    .split('\n')
                    .collect()
            ),
            2
        );

        assert_eq!(
            process(
                r"LLR

AAA = (BBB, BBB)
BBB = (AAA, ZZZ)
ZZZ = (ZZZ, ZZZ)"
                    .split('\n')
                    .collect()
            ),
            6
        );
    }
}

File: 2023/java/d08p1.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;

public class d08p1 {
    public static void main(String[] args) throws IOException {
        System.out.println("Day 8 - Part 1");

        BufferedReader reader = new BufferedReader(new FileReader("../_input/day08.txt"));

        long total = process(reader);
        reader.close();

        System.out.println("Total: " + total);
    }

    private static int process(BufferedReader reader) throws IOException {
        Direction[] instructions = null;
        HashMap<String, Node> nodes = new HashMap<>();

        String line;
        while ((line = reader.readLine()) != null) {
            if (instructions == null) instructions = processDirections(line);
            else if (!line.isBlank()) {
                var node = processNode(line);
                nodes.put(node.name, node);
            }
        }

        String currentNode = "AAA";
        int currentDir = 0;
        boolean hit = false;
        int counter = 0;
        while (!hit) {
            Node node = nodes.get(currentNode);
            switch (instructions[currentDir]) {
                case Direction.LEFT -> currentNode = node.left;
                case Direction.RIGHT -> currentNode = node.right;
            }

            counter++;
            if (currentNode.equals("ZZZ")) hit = true;

            currentDir++;
            if (currentDir >= instructions.length) currentDir = 0;
        }

        return counter;
    }

    private static Direction[] processDirections(String line) {
        ArrayList<Direction> dirs = new ArrayList<>();
        for (int c : line.chars().toArray()) dirs.add(Direction.get(c));
        return dirs.toArray(Direction[]::new);
    }

    private static Node processNode(String line) {
        String[] n = line.split(" = ");
        String[] ns = n[1].substring(1, n[1].length() - 1).split(", ");
        return new Node(n[0], ns[0], ns[1]);
    }

    private enum Direction {
        LEFT,
        RIGHT;

        public static Direction get(int c) {
            switch (c) {
                case (int) 'L' -> {
                    return Direction.LEFT;
                }
                case (int) 'R' -> {
                    return Direction.RIGHT;
                }
                default -> throw new RuntimeException("Not a direction.");
            }
        }
    }

    private record Node(String name, String left, String right) {}
}
